keep/step/step.py

```diff
@@ -167,28 +172,34 @@ def _run_single(self):
             result = self._run_single_async()
         # else, just run the provider
         else:
-            try:
-                rendered_providers_parameters = {}
-                for parameter in self.provider_parameters:
-                    rendered_providers_parameters[parameter] = self.io_handler.render(
-                        self.provider_parameters[parameter]
+            for curr_retry_count in range(self.__retry_count + 1):
+                try:
+                    rendered_providers_parameters = {}
+                    for parameter in self.provider_parameters:
+                        rendered_providers_parameters[parameter] = self.io_handler.render(
+                            self.provider_parameters[parameter]
+                        )
```

Comment:
This can be taken out of the for loop, there's no need to re-render the parameters every time.